alchemiscale_fah/compute/service.py

```diff
@@ -4,6 +4,621 @@
+
+        # execute the task; this looks the same whether the ProtocolDAG is a
+        # success or failure
+
+        shared = self.shared_basedir / str(protocoldag.key)
+        shared.mkdir()
+        scratch = self.scratch_basedir / str(protocoldag.key)
+        scratch.mkdir()
+
+        self.logger.info("Executing '%s'...", protocoldag)
+        try:
+            # use a custom `execute_DAG` here that feeds appropriate components
+            # via context, such as the FahAdaptiveSamplingClient, to units that
+            # interact with FAH
+            protocoldagresult = await execute_DAG(
+                protocoldag,
+                shared_basedir=shared,
+                scratch_basedir=scratch,
+                keep_scratch=self.keep_scratch,
+                raise_error=False,
+                n_retries=self.settings.n_retries,
+                pool=self._pool,
+                fah_client=self.fah_client,
+                fah_projects=self.fah_projects,
+                project_run_clone=(project_id, run_id, clone_id),
+                transformation_sk=tf_sk,
+                task_sk=task,
+                index=self.index,
+            )
+        finally:
+            if not self.keep_shared:
+                shutil.rmtree(shared)
+
+            if not self.keep_scratch:
+                shutil.rmtree(scratch)
+
+        if protocoldagresult.ok():
+            self.logger.info("'%s' -> '%s' : SUCCESS", protocoldag, protocoldagresult)
+        else:
+            for failure in protocoldagresult.protocol_unit_failures:
+                self.logger.info(
+                    "'%s' -> '%s' : FAILURE :: '%s' : %s",
+                    protocoldag,
+                    protocoldagresult,
+                    failure,
+                    failure.exception,
+                )
+
+        # push the result (or failure) back to the compute API
+        result_sk = self.push_result(task, protocoldagresult)
+        self.logger.info("Pushed result `%s'", protocoldagresult)
+
+        return task, result_sk
+
+    async def async_cycle(
+        self, max_tasks: Optional[int] = None, max_time: Optional[int] = None
+    ):
+        self._check_max_tasks(max_tasks)
+        self._check_max_time(max_time)
+
+        # claim as many tasks as we are allowed to at once
+        # TODO: only want to claim tasks that correspond to FAH protocols
+        # claim tasks from the compute API
+        self.logger.info("Claiming tasks")
+        task_sks: List[ScopedKey] = self.client.claim_tasks(
+            scopes=self.scopes,
+            compute_service_id=self.compute_service_id,
+            count=self.claim_limit,
+            protocols=self.settings.protocols,
+        )
+
+        # if no tasks claimed, sleep and return
+        if all([task_sk is None for task_sk in task_sks]):
+            self.logger.info(
+                "No tasks claimed; sleeping for %d seconds", self.sleep_interval
```

Comment:
Fstring here and below

Reply:
up to you ofc, but IMO better to be modern and consistent.

Reply:
For logging, still considered more efficient to use `%s` usage: https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs